clients/src/main/java/org/apache/kafka/clients/producer/internals/Sender.java

```diff
@@ -685,6 +711,13 @@ private void completeBatch(ProducerBatch batch, ProduceResponse.PartitionRespons
                             "to request metadata update now", batch.topicPartition,
                             error.exception(response.errorMessage).toString());
                 }
+                if (error.exception() instanceof NotLeaderOrFollowerException || error.exception() instanceof FencedLeaderEpochException) {
+                    log.debug("For {}, received error {}, with leaderIdAndEpoch {}", batch.topicPartition, error, response.currentLeader);
+                    if (partitionsWithUpdatedLeaderInfo != null && response.currentLeader.leaderId() != -1 && response.currentLeader.leaderEpoch() != -1) {
```

Comment:
This line could be more legible with some parentheses and splitting the condition across multiple lines.